hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.spi.state;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class WritableKVStateBase<K extends Comparable<K>, V>
+        extends ReadableKVStateBase<K, V> implements WritableKVState<K, V> {
+    /** A map of all modified values buffered in this mutable state */
+    private final Map<K, V> modifications = new HashMap<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    protected WritableKVStateBase(@NonNull final String stateKey) {
+        super(stateKey);
+    }
+
+    /**
+     * Flushes all changes into the underlying data store. This method should <strong>ONLY</strong>
+     * be called by the code that created the {@link WritableKVStateBase} instance or owns it. Don't
+     * cast and commit unless you own the instance!
+     */
+    public final void commit() {
+        for (final var entry : modifications.entrySet()) {
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (value == null) {
+                removeFromDataSource(key);
+            } else {
+                putIntoDataSource(key, value);
```

Comment:
If the backing data source is a `MerkleMap`, I'm 95% sure we have to ensure a deterministic order of `remove()`'s and `put()`'s of new leaves (@cody-littley can you confirm?) 

Since it would be a shame to have to sort in the 99% of cases where there _are_ no leaves added or removed, we may also want to consider specialized `insert()` and `remove()` methods with separate data structures tracking created, removed, and modified keys.

Reply:
Oh! If I understand correctly, you're saying that every node must call put and remove for keys in exactly the same sequence (whichever sequence it happens to be). For example, if I have to remove A and C and put B, then it doesn't matter whether we remove A then C then put B, or if we remove A then put B then remove C -- AS LONG AS all nodes do it in exactly the same order.

You are right, this must be deterministic across all nodes, even with VirtualMap, or it will create different trees.

Even if we had separate data structures for added, removed, and modified items, we'd still have to sort added and removed so they are done deterministically (which I think you're saying), but at least the modified items could be unsorted.

Really good catch!!

Reply:
Would a `LinkedHashMap` for `modifications` solve the problem?

Reply:
I wrote a "hammer" test to try to catch indeterministic behavior, and I was not able to get it, even with 100,000 iterations of 100 random mutations in 39 different MutableKVState instances modified each by a different thread. This seems to imply that a HashMap is deterministic if modified in the same order.

Now, most of our KV states will have something on the order of 1-5 modifications, and a HashMap may be overkill for such a thing anyway. Maybe I should change the implementation to be just a list, and when we need to look up the key (contains, get, etc) we just iterate over the list. It is likely faster than using map, and would use substantially less memory, and would guarantee consistent ordering. I'm not sure if this is a block or not, given that the test does not fail.

Reply:
I would have been surprised, if the iteration order of `HashMap` was non-deterministic. 🙂 But it would still be wrong to rely on this IMO. As the order is not guaranteed, two different Java versions could have deterministic, but differing iteration orders. I.e. if two nodes run different Java versions, they may end up with different results.

I agree, if there are just a handful of modifications, a `HashMap` is overkill.

Reply:
Even after adding a bunch of random junk into the map and then clearing it, it is STILL deterministic! This test passes!

```
void deterministicUpdates() {
        // We will execute 'numIterations' iterations. This is parameterized with a system property
        // so that we may dial the number up or down for certain types of tests (for example,
        // nightly vs. continuous). Each iteration will construct a list of modifications (add,
        // remove, modify). We will then create 39 threads each with their own WritableKVStateBase.
        // Each will apply the modifications and commit them. We will then join back and verify
        // that the order in which modifications were applied is identical. The hope is that if
        // there is any non-determinism, we will tease it out by hammering it in this way.
        final int numIterations = Integer.getInteger(NUM_ITERATIONS_ARG, 100);
        final var numMutations = 20;
        final var numThreads = 39;
        final var executors = Executors.newFixedThreadPool(numThreads);
        for (int i = 0; i < numIterations; i++) {
            // Contains lambdas that will mutate the map in some way, and then populate the
            // mutation list with random mutations (puts and removes)
            List<Consumer<WritableKVStateBase<Integer, String>>> mutations = new ArrayList<>();
            for (int j = 0; j < numMutations; j++) {
                final var key = random().nextInt(100);
                if (random().nextInt(10) < 8) {
                    final var randomSuffix = randomString(10);
                    mutations.add(state -> state.put(key, randomSuffix));
                } else {
                    mutations.add(state -> state.remove(key));
                }
            }

            // Now, spin up the threads. Each one will have its own unique state. The state
            // is special in that it keeps track of the order in which remove and add calls
            // were made to it, so we can compare those later.
            final var latch = new CountDownLatch(numThreads);
            final var mutationOrders = new ArrayList<List<Integer>>();
            for (int t = 0; t < numThreads; t++) {
                final var state =
                        new MapWritableKVState<Integer, String>(FRUIT_STATE_KEY) {
                            private final List<Integer> keys = new ArrayList<>();

                            @Override
                            protected void putIntoDataSource(
                                    @NonNull Integer key, @NonNull String value) {
                                keys.add(key);
                                super.putIntoDataSource(key, value);
                            }

                            @Override
                            protected void removeFromDataSource(@NonNull Integer key) {
                                keys.add(key);
                                super.removeFromDataSource(key);
                            }
                        };

                // Add a bunch of random junk to the state and then reset it.
                // If the backend data structures are not able to maintain deterministic
                // behavior in the face of this (like a HashMap), then the test will fail.
                for (int j = 0; j < random().nextInt(1000, 100_000_000); j++) {
                    state.put(random().nextInt(), randomString(100));
                }
                state.reset();

                mutationOrders.add(state.keys);
                executors.execute(
                        () -> {
                            for (var mutator : mutations) {
                                mutator.accept(state);
                            }
                            state.commit();
                            latch.countDown();
                        });
            }

            try {
                if (!latch.await(1, TimeUnit.SECONDS)) {
                    fail("Failed to complete task in a reasonable time");
                }
            } catch (InterruptedException e) {
                fail("Test was interrupted");
            }

            // Now we verify that every single "mutationOrder" array is the same
            List<Integer> mutationOrder = mutationOrders.get(0);
            for (final var mo : mutationOrders) {
                assertThat(mo).containsExactlyElementsOf(mutationOrder);
            }
        }
    }
```

Reply:
I just switch from HashMap to LinkedHashMap to be safe, and I also wrote a quick JMH benchmark showing on my mac at least 10,000,000 ops/sec, where each op was 10 modifications to the map. I think that is plenty fast for now and won't be a bottleneck, so I think we're good to ship with LinkedHashMap for now, unless we find we need to do something more intense due to performance testing and analysis